src/js/node/net.ts

```diff
@@ -667,264 +1122,289 @@ Socket.prototype[kCloseRawConnection] = function () {
+                const [result, events] = upgradeDuplexToTLS(connection, {
+                  data: { self: this, that: socket, req: { oncomplete: afterConnect } },
+                  tls,
+                  socket: this[khandlers],
+                });
+                connection.on("data", events[0]);
+                connection.on("end", events[1]);
+                connection.on("drain", events[2]);
+                connection.on("close", events[3]);
+                this._handle = SocketHandle[kAttach](result, this);
               } else {
-                this._handle = null;
-                throw new Error("Invalid socket");
+                this.connecting = true;
+                this[kupgraded] = connection;
+                connection.connecting = true;
+                const result = socket.upgradeTLS({
+                  data: { self: this, that: socket, req: { oncomplete: afterConnect } },
+                  tls,
+                  socket: this[khandlers],
+                });
+                if (result) {
+                  const [raw, tls] = result;
+                  // replace socket
+                  connection._handle = SocketHandle[kAttach](raw, connection);
+                  this.once("end", this[kCloseRawConnection]);
+                  raw.connecting = false;
+                  this._handle = SocketHandle[kAttach](tls, this);
+                } else {
+                  this._handle = null;
+                  throw new Error("Invalid socket");
+                }
               }
-            }
-          });
+            });
+          }
         }
+      } catch (error) {
+        process.nextTick(emitErrorAndCloseNextTick, this, error);
       }
-    } else if (path) {
-      // start using unix socket
-      bunConnect({
-        data: this,
-        unix: path,
-        socket: this[khandlers],
-        tls,
-        allowHalfOpen: this.allowHalfOpen,
-      }).catch(error => {
-        if (!this.destroyed) {
-          this.emit("error", error);
-          this.emit("close");
-        }
-      });
-    } else {
-      // default start
-      bunConnect({
-        data: this,
-        hostname: host || "localhost",
-        port: port,
-        socket: this[khandlers],
-        tls,
-        allowHalfOpen: this.allowHalfOpen,
-      }).catch(error => {
-        if (!this.destroyed) {
-          this.emit("error", error);
-          this.emit("close");
-        }
-      });
+      return this;
     }
-  } catch (error) {
-    process.nextTick(emitErrorAndCloseNextTick, this, error);
+  }
+
+  const [options, cb] = $isArray(args[0]) && args[0][normalizedArgsSymbol] ? args[0] : normalizeArgs(args);
+  const connectListener = cb;
+
+  if (typeof this[bunTlsSymbol] === "function" && cb !== null) {
+    this.once("secureConnect", cb);
+  } else if (cb !== null) {
+    this.once("connect", cb);
+  }
+  if (this._parent?.connecting) {
+    return this;
+  }
+  if (this.write !== Socket.prototype.write) {
```

Comment:
This is suspicious

Reply:
yes https://github.com/nodejs/node/blob/v24.0.0/lib/net.js#L1254